src/digraph.rs

```diff
@@ -2265,6 +2266,150 @@ impl PyDiGraph {
+    ///     that the substitution will not introduce cycles before
+    ///     modifying the graph, even if cycle check is disabled for
+    ///     this instance of :class:`retworkx.PyDiGraph`. Otherwise,
+    ///     the value of member ``check_cycle`` is inherited.
+    /// :raises DAGWouldCycle: The cycle check is enabled and the
+    ///     substitution would introduce cycle(s).
+    /// :raises ValueError: ``to_replace`` is empty or ``node``
+    ///     has existing connections.
+    #[pyo3(
+        text_signature = "(self, to_replace, node, /, force_check_cycle=None)"
+    )]
+    fn substitute_nodes_with_node(
+        &mut self,
+        py: Python,
+        to_replace: HashSet<usize>,
+        node: usize,
+        force_check_cycle: Option<bool>,
+    ) -> PyResult<()> {
+        fn can_contract(
+            graph: &StablePyGraph<Directed>,
+            nodes: &HashSet<NodeIndex>,
+        ) -> bool {
+            let mut visited = graph.visit_map();
+            let mut visit_next = VecDeque::new();
+
+            // Start with successors of `nodes` that aren't in `nodes` itself.
+            for node in nodes {
+                for edge in graph.edges(*node) {
+                    if !nodes.contains(&edge.target()) {
+                        visit_next.push_back(edge.target());
+                    }
+                }
+            }
+
+            // DFS. If we encounter any of `nodes`, there exists a path from `nodes`
+            // back to `nodes` of length > 1, meaning contraction is disallowed.
+            while let Some(decendant) = visit_next.pop_back() {
+                if visited.is_visited(&decendant) {
+                    continue;
+                }
+
+                for edge in graph.edges(decendant) {
+                    if nodes.contains(&edge.target()) {
+                        // we found a path back to `nodes`
+                        return false;
+                    }
+
+                    visit_next.push_back(edge.target());
+                }
+
+                visited.visit(decendant);
+            }
+
+            true
+        }
+
+        macro_rules! assert_node_exists {
+            ($node_index:ident) => {
+                if self.graph.node_weight($node_index).is_none() {
+                    return Err(PyIndexError::new_err(format!(
+                        "Specified node {} is not in this graph",
+                        $node_index.index()
+                    )));
+                }
+            };
+        }
+
+        if to_replace.is_empty() {
+            return Err(PyValueError::new_err("to_replace must not be empty"));
+        }
+
+        let node_index = NodeIndex::new(node);
+        assert_node_exists!(node_index);
+
+        if self.graph.neighbors_undirected(node_index).next().is_some() {
+            return Err(PyValueError::new_err(
+                "replacement node must not have connections",
+            ));
+        }
```

Comment:
If `node` is an existing node, we should allow it to have edges and adjust the `can_contract` cycle check. Another option would be to take as input a weight `PyObject`, insert a new node in the graph with the given weight and return back the new index.

Reply:
I'll agree with Georgios and say that taking an optional weight and returning a new node index seems like a better design choice. Creating an empty node with `add_node` and then calling this function is not as convenient.

Reply:
Agreed. Done in c8699ae.